Auto-formatting tweaks

pugnascotia opened the pull request:
Follow-up to #46745. Tweak the Eclipse formatter config. It's hard to see what different is from the XML, so here's a summary:

   * Split method, constructor etc arguments into separate lines when they wrap
   * Enable "joined wrapped lines", which basically means the formatter will ignore existing line breaks and join and re-wrap statements. I had originally hesitated about enabling this, but without it the codebase's style won't be as consistent as it should be, defeating the point of all this formatting business.

Also tweak some source files so that they aren't adversely affected when we finally format the source.

   * Convert some ASCII art diagrams that were in line quotes into to block quotes, which the formatter is configured to ignore.
   * Apply some formatter off/on comments to shield some blocks. There are just a few places I happened to notice.

pugnascotia committed:
Change comment style to preserve ASCII art

pugnascotia committed:
Proof some code against the formatter

pugnascotia committed:
Refine code formatter settings

elasticmachine commented:
Pinging @elastic/es-core-infra

rjernst commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/SslConfigurationLoader.java:
```diff
@@ -68,6 +68,7 @@
 
     static final List<String> DEFAULT_PROTOCOLS = List.of("TLSv1.3", "TLSv1.2", "TLSv1.1");
 
+    // @formatter:off
```
This seems like an important tool we should document in our CONTRIBUTING.md? (separate from this PR)

pugnascotia replied:
I can add it to `CONTRIBUTING.md`, I should have adding it in the last PR to be honest.

rjernst commented on server/src/main/java/org/apache/lucene/queries/BinaryDocValuesRangeQuery.java:
```diff
@@ -145,25 +145,31 @@ public int hashCode() {
-                // part of the other range must touch this range
-                // this:    |---------------|
-                // other:               |------|
+                /*
```
Why is this converted to a multiline comment? Is there some problem with using several single line comments with the formatter?

pugnascotia replied:
I observed the formatted mangling these diagrams, hence the change. You can ask the formatter to process block and javadoc comments, but I decided that wasn't worth it.

rjernst commented on server/src/test/java/org/elasticsearch/search/suggest/SuggestionTests.java:
```diff
@@ -196,17 +198,20 @@ public void testToXContent() throws IOException {
-                                + "\"collate_match\":true}]"
-                            + "}]"
-                    + "}", xContent.utf8ToString());
+                // @formatter:off
```
We have lines like this in a lot of tests. Is there anything we can do to improve this without resorting to manually turning off the formatter?

pugnascotia replied:
There are options, depending on the code in question. 

### JSON snippets

For this example, we could include whitespace:

```java
String suggestionString = "{"
    + "\"unknownType#suggestionName\": ["
    + "  {"
    + "    \"text\": \"entryText\","
    + "    \"offset\": 42,"
    + "    \"length\": 313,"
    + "    \"options\": [{"
    + "      \"text\": \"someText\","
    + "      \"highlighted\": \"somethingHighlighted\","
    + "      \"score\": 1.3,"
    + "      \"collate_match\": true"
    + "    }]"
    + "  }]"
    + "}";
```

With Java 13, we could use [text blocks](https://openjdk.java.net/jeps/355):

```
String suggestionString = """
    {
    "unknownType#suggestionName": [
      {
        "text": "entryText",
        "offset": 42,
        "length": 313,
        "options": [{
          "text": "someText",
          "highlighted": "somethingHighlighted",
          "score": 1.3,
          "collate_match": true
        }]
      }]
    }
    """;
```

We could also make these snippets more readable by using an `ObjectMapper` with some lenient settings, e.g:

```java
public static String asJson(final String rawJson) throws IOException {
    ObjectMapper mapper = new ObjectMapper();

    mapper.configure(JsonParser.Feature.ALLOW_UNQUOTED_FIELD_NAMES, true);
    mapper.configure(JsonParser.Feature.ALLOW_SINGLE_QUOTES, true);

    final JsonNode jsonNode = mapper.readTree(rawJson);

    return mapper.writeValueAsString(jsonNode);
}
```

Then we can write the JSON like so:

```java
String suggestionString = asJson("{"
    + "'unknownType#suggestionName': ["
    + "  {"
    + "    text: 'entryText',"
    + "    offset: 42,"
    + "    length: 313,"
    + "    options: [{"
    + "      text: 'someText',"
    + "      highlighted: 'somethingHighlighted',"
    + "      score: 1.3,"
    + "      collate_match: true"
    + "    }]"
    + "  }]"
    + "}");
```

Depend how wild we're feeling, we could define these snippets in YAML, and convert to JSON, e.g.:

```java
String suggestionYaml = yamlToJson(
        "unknownType#suggestionName:\n"
        + "  - text: entryText\n"
        + "    offset: 42\n"
        + "    length: 313\n"
        + "    options:\n"
        + "      - text: someText\n"
        + "        highlighted: somethingHighlighted\n"
        + "        score: 1.3\n"
        + "        collate_match: true\n");
```

Though I expect that'll look better with text blocks, once they're available:

```
String suggestionYaml = yamlToJson("""
    unknownType#suggestionName:
      - text: entryText
        offset: 42
        length: 313
        options:
          - text: someText
            highlighted: somethingHighlighted
            score: 1.3
            collate_match: true
    """);
```

### Aligned code

The changes to [SslConfigurationLoader.java](https://github.com/elastic/elasticsearch/pull/47299/files#diff-e479ea9b1c908fc4e30392ff15c413c1) are a case where strings and comments are arranged in meaningful rows. I think this is a case where just exempting the code is the best approach.

On that subject, however, the Painless module is interesting because it has code that has vertical alignment. For example:

   - [DefMath.java](https://github.com/elastic/elasticsearch/blob/master/modules/lang-painless/src/main/java/org/elasticsearch/painless/DefMath.java#L1083)
   - [MethodWriter.java](https://github.com/elastic/elasticsearch/blob/master/modules/lang-painless/src/main/java/org/elasticsearch/painless/MethodWriter.java#L184)
   - [Operation.java](https://github.com/elastic/elasticsearch/blob/master/modules/lang-painless/src/main/java/org/elasticsearch/painless/Operation.java#L30)
   - [WriterConstants.java](https://github.com/elastic/elasticsearch/blob/master/modules/lang-painless/src/main/java/org/elasticsearch/painless/WriterConstants.java#L137)

I can't decide whether we should turn off the formatter for these. What do you think?

Also, Spotless doesn't appear to have a way to exempt files from formatting, so it formats all the source that ANTLR generates too. Can't do much about that.

pugnascotia replied:
@rjernst do you want me to do anything here? I could go looking for more code whose format we want to preserve?

rjernst replied:
Long term I think we will want to use text blocks. I wouldn't change to yaml, since we are testing json. Ideally these would actually be converted to xcontentbuilder, since that has earlier error checking (eg invalid json would get caught when running the code, rather than in the request parsing the json). My concern here is starting this precedent of arbitrarily turning off the formatter. Using column aligned strings seems like the best option for now, but we should probably change those in separate PRs from any enabling auto formatting, so that the formatting PRs can remain completely automatic and mechanical in nature, thus simplifying reviewing.

rjernst reviewed (commented):
I left some comments

pugnascotia committed:
Note how to disable formatting in CONTRIBUTING.md

pugnascotia committed:
Fix missing backtick in CONTRIBUTING.md

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into formatting-tweaks

pugnascotia committed:
Better whitespace formatting, fix tag/end handling

For cases where the formatter was disabled around JSON snippets, embed
more whitespace within the strings to as to format the JSON in a way
that the formatter won't break.

Also, change the formatter's enable/disables tags to that the
documentation comments i.e "// tag::NAME" and "// end::NAME" still work.
This means that such snippets won't be automatically formatted, but
since the formatter can't be configured different for specific regions
in a file, it's the best we can do.

pugnascotia committed:
Various formatting improvements

Make a number of code changes so that when the codebase is automatically
formatted, the code doens't end up looking mangled. Mostly this is:

- Moving comments
- Joining multiline strings that fit on one line
- Formatting JSON so that indentation is inside the strings
  - This includes stripping out additional whitespace in some cases,
    where the test does an exact comparison with JSON from the API

Also reformat XML documents in the SAML tests, using a formatter that
can handle named parameters. This significantly improves legibility.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into formatting-tweaks

pugnascotia commented:
@rjernst I've reworked the code where I'd disabled the formatter, and then carried on to audit the repository post-formatting. I've made a number of further changes so that some code doesn't appear mangled after formatting. Mostly this is:

- Moving comments around
- Joining multiline strings that fit on one line (I didn't do all these, there were a lot)
- Formatting JSON so that indentation is inside the strings
  - This includes stripping out additional whitespace in some cases,
    where the test does an exact comparison with JSON from the API. I added a method to `XContentHelper` for this.

I also reformatted the XML documents in the SAML tests, using a formatter that
can handle named parameters. This significantly improved legibility.

There are still some other files that could use some love, for example the geo parser tests. They're building up documents via XContent, and since the formatter sees these as a long sequence of method calls, that exactly how they're formatted. I think we're better off just using JSON here - it'd be significantly terser, and would retain indentation.

pugnascotia commented:
@elasticmachine test this please

pugnascotia committed:
Test fix

pugnascotia committed:
Put NamedFormatter in a better location

pugnascotia committed:
Fix forbidden APIs

pugnascotia committed:
Checkstyle exception

pugnascotia committed:
Linting

pugnascotia committed:
Test fixes

pugnascotia committed:
More tests and checkstyle

pugnascotia committed:
Lint

rjernst commented:
Thanks for working through the formatter disabling issues.

Something I've realized upon looking back at this is the enormous size in conjunction with several changes being made to the format. Would it be possible to separate out each of the listed tweaks in the description to their own PRs? If a PR is 1k lines, it is generally difficult to adequately review, and this PR is almost 4k edits. I understand these tweaks will necessarily touch a lot of code and be large, but it is much easier to scroll through a large PR when there is a singular purpose, and thus one typical pattern to visually look for.

pugnascotia commented:
@rjernst sure, the PR wasn't really supposed to be this size. I'll split up the changes and list them here for posterity.

   * [ ] Refine the auto-formatting settings (#48446)
